python/test_dcr_template.py

```diff
@@ -0,0 +1,629 @@
+        model_vals = DcrModel._extract_model_vals(j_use, i_use, radius=radius, model=self.dcrModel.model,
+                                                  weights=self.dcrModel.weights)
+        dcr_kernel = DcrModel._calc_offset_phase(self.exposure, self.dcr_gen, return_matrix=True,
+                                                 x_size=self.kernel_size, y_size=self.kernel_size)
+        dcr_vals = DcrModel._apply_dcr_kernel(dcr_kernel, model_vals)
+        dcr_ref = np.load(data_file)
+        self.assertFloatsEqual(dcr_vals, dcr_ref)
+
+    @unittest.expectedFailure
+    def test_apply_even_kernel(self):
+        """Only odd kernel sizes are currently supported, so this test is expected to fail for now."""
+        data_file = "test_data/dcr_kernel_even_vals.npy"
+        kernel_size = 6
+        i_use = self.size//2
+        j_use = self.size//2
+        radius = kernel_size//2
+        model_vals = DcrModel._extract_model_vals(j_use, i_use, radius=radius, model=self.dcrModel.model,
+                                                  weights=self.dcrModel.weights)
+        dcr_kernel = DcrModel._calc_offset_phase(self.exposure, self.dcr_gen, return_matrix=True,
+                                                 x_size=kernel_size, y_size=kernel_size)
+        dcr_vals = DcrModel._apply_dcr_kernel(dcr_kernel, model_vals)
+        np.save(data_file, dcr_vals)
+        dcr_ref = np.load(data_file)
+        self.assertFloatsEqual(dcr_vals, dcr_ref)
+
+
+class PersistanceTestCase(DcrModelTestBase):
+    """Tests that read and write exposures and dcr models to disk."""
+
+    def test_create_exposure(self):
+        self.assertFloatsEqual(self.exposure.getMaskedImage().getImage().getArray(), self.array)
+        meta = self.exposure.getMetadata()
+        # Check that the required metadata is present:
+        try:
+            meta.get("ZENITH")
+        except Exception as e:
+            raise e
+        try:
+            meta.get("AZIMUTH")
+        except Exception as e:
+            raise e
+
+    def test_persist_dcr_model_roundtrip(self):
+        # Instantiating the butler takes several seconds, so all butler-related tests are condensed into one.
+        model_repository = "test_data"
+
+        # First test that the model values are not changed from what is expected
+
+        # The type "dcrModel" is read in as a 32 bit float, set in the lsst.obs.lsstSim.LsstSimMapper policy
+        model = np.float32(self.dcrModel.model)
+        self.dcrModel.export_model(model_repository=model_repository)
+        dcrModel2 = DcrModel(model_repository=model_repository)
+        # Note that butler.get() reads the FITS file in 32 bit precision.
+        self.assertFloatsEqual(model, dcrModel2.model)
+
+        # Next, test that the required parameters have been restored
+        param_ref = self.dcrModel.__dict__
+        param_new = dcrModel2.__dict__
+        for key in param_ref.keys():
+            self.assertIn(key, param_new)
```

Comment:
Can't you just compare the keys from each dict, without a loop? Or even just compare the dicts for equality?